### PR TITLE
logseq: add darwin support

### LINKED ISSUE #258818: Package request: Logseq (for mac
**Project description**

**Metadata**

* homepage URL:
* source URL:
* license: mit, bsd, gpl2+ , ...
* platforms: unix, linux, darwin, ...

### PR DESCRIPTION
**Pull request by yzx9**
## Description of changes

This PR adds support for Logseq on Darwin, although there are still a few unresolved issues:

1. An alert stating "This file was downloaded on an unknown date." prevents the app from opening. This can be resolved by allowing it in Settings > Privacy & Security > Allow. The cause of this issue is unclear, but it may be related to #320494.
2. Unlike the Linux version, I am unsure how to remove the internal git and sign the app again.

Additionally, there is an issue (#296939) tracking the requirement for Electron apps to be built from source. I attempted to build it from source, which involves converting ClojureScript to JavaScript, but encountered problems with [clj2nix](https://github.com/hlolli/clj2nix/) due to [clj2nix#12](https://github.com/hlolli/clj2nix/issues/12). Furthermore, [clj-nix](https://github.com/jlesquembre/clj-nix) is not currently upstreamed ([clj-nix#72](https://github.com/jlesquembre/clj-nix/issues/72)).

Some issues remain unaddressed, and I am drafting this PR to look for your help. I would greatly appreciate any solutions, examples for creating Clojure applications, or packaging samples.

Thank you in advance for any help or additional information.

Closes #258818

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by superherointj** [comment]
Please fix commit message title (darwin).

File: pkgs/by-name/lo/logseq/package.nix [hunk 1]
```diff
@@ -13,32 +14,43 @@
...
 
 stdenv.mkDerivation (finalAttrs: let
-  inherit (finalAttrs) pname version src appimageContents;
-
+  inherit (finalAttrs) pname version src;
+  inherit (stdenv.hostPlatform) system;
+  selectSystem = attrs: attrs.${system} or (throw "Unsupported system: ${system}");
+  suffix = selectSystem {
+    x86_64-linux = "linux-x64-${version}.AppImage";
+    x86_64-darwin = "drawin-x64-${version}.zip";
```
Comment (superherointj) [inline pkgs/by-name/lo/logseq/package.nix:22]:
```suggestion
    x86_64-darwin = "darwin-x64-${version}.zip";
```

Is `drawin` on purpose?

Comment (yzx9) [reply pkgs/by-name/lo/logseq/package.nix]: My typo, which I keep making over and over, especially when I'm sleepy ;)

**Comment by superherointj** [comment]
Result of `nixpkgs-review pr 331824` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>logseq</li>
  </ul>
</details>

**Comment by superherointj** [comment]
Result of `nixpkgs-review pr 331824` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package marked as broken and skipped:</summary>
  <ul>
    <li>logseq</li>
  </ul>
</details>

**Comment by yzx9** [comment]
I just realized there are three PRs (#191091, #248090, #268249) that have attempted to add Darwin support. However, I have noticed that they have all been marked as stale for more than half a year.

Unlike other PRs which separate the derivations between platforms, this PR aims to minimize changes, making it clearer and easier to review. Therefore, I still hope this can be merged.

I am also willing to assist in merging the other three PRs. Hope we can see Darwin support for logseq soon.

@kirillrdy @klchen0112, could I ask you to review and possibly help move this PR forward? 



---

Regarding some comments from @klchen0112 on https://github.com/NixOS/nixpkgs/pull/248090#issuecomment-1912065616

> 1. It's necessary to use makeWrapper in darwin? #268249

I think it's okay, more of a habit. However, if you prefer to remove it, I am willing to do so.


> 2. For linux, I think this is enough. #248090 add `curl expat openssl zlib` in buildinputs
>
> ```
> nativeBuildInputs = [ makeWrapper autoPatchelfHook ];
> buildInputs = [ stdenv.cc.cc];
> ```

I agree with you that `curl expat openssl zlib` is not needed in #191091.


> 3. `runHook preInstall` and `runHook postInstall` should be added for darwin build?

Haha, I just know it is optional. I think maybe it is helpful for overlays?

**Comment by kirillrdy** [comment]
@yzx9 thank you for perseverance, I would like this resolved as well.

currently logseq is broken ( on linux) due to 

```
  Known issues:
        - Electron version 27.3.11 is EOL
```

we would need to resolve this first ( somehow )

**Comment by kirillrdy** [comment]
actually maybe fixing linux might not be relevant, let me check a few things

**Comment by yzx9** [comment]
@kirillrdy Hey, any updates? I suggest we prioritize adding Darwin support first. The platform support for Linux and Darwin is actually independent. That said, building Logseq from source is not a straightforward task, as I mentioned earlier.

**Review by kirillrdy — APPROVED** [review]
image

linux build also still works ( just need to allow insecure )

**Pull request merged**